### PR TITLE
Fix importing of translated csv dynamic values

### LINKED ISSUE #321: Make sure dynamic strings are not changed when importing translations
Sometimes translators will change `{{price}}` to `{{precio}}`. We should make sure values inside `{{ }}` are present in both strings, and unchanged.

### PR DESCRIPTION
**Pull request by DeeTheDev**
Fixes #321 

- Add method to ensure dynamic values do not change during translated CSV import
- Test

### CODE REVIEW

File: spec/suma/i18n_spec.rb [hunk 1]
```diff
@@ -58,6 +58,36 @@
     end
   end
 
+  describe "ensure_english_interpolation_values" do
```
Comment (rgalanakis) [inline spec/suma/i18n_spec.rb:61]: These tests need to go 'through the front door' and use `import_csv`. If that's hard, you can refactor `import_csv` to pull out a high-level, easier-to-test method.

Try to avoid testing implementation details like this explicitly. In some cases, it's fine for edge cases that are hard to exercise through the front door. But in general, you want to avoid tests that break for internal refactorings (and this is definitely an internal thing, no one would need to call it directly). For one benefit- you don't ever test this code is actually used :)

File: spec/suma/i18n_spec.rb [hunk 2]
```diff
@@ -58,6 +58,36 @@
...
+      english_dynamic_str = "Price is {{price}}."
+      expect(described_class.ensure_english_interpolation_values(
+               spanish_dynamic_str,
+               english_dynamic_str,
+               "English",
+             )).to eq("Precio es {{price}}.")
+    end
+    it "ensures multiple dynamic values are not changed" do
+      spanish_dynamic_str = "Precio es {{precio}}. Salvaste {{cantidad}}."
+      english_dynamic_str = "Price is {{price}}. You saved {{amount}}."
```
Comment (rgalanakis) [inline spec/suma/i18n_spec.rb:73]: What happens if the ordering of translated strings is reversed? Ie, `{{price}} {{amount}}` and `{{cantidad}} {{precio}}`.

File: lib/suma/i18n.rb [hunk 3]
```diff
@@ -125,6 +125,30 @@ def self.flatten_hash(h, memo: {}, path: [])
...
   end
 
+  # Ensures a string with interpolation/dynamic value (surrounded by
+  # curly brackets) is not changed to a different value, for example:
+  # A Spanish string with interpolation value `{{precio}}` would be reset
+  # to `{{price}}`.
+  # @return [String]
+  def self.ensure_english_interpolation_values(str, other_str, lng="English")
+    return str unless str.include?("{{") && other_str.include?("{{") && lng === "English"
+    if str.scan("{{").length === 1
```
Comment (rgalanakis) [inline lib/suma/i18n.rb:135]: Do we need this special case for one interpolation?

File: lib/suma/i18n.rb [hunk 4]
```diff
@@ -125,6 +125,30 @@ def self.flatten_hash(h, memo: {}, path: [])
     return memo
   end
 
+  # Ensures a string with interpolation/dynamic value (surrounded by
+  # curly brackets) is not changed to a different value, for example:
+  # A Spanish string with interpolation value `{{precio}}` would be reset
+  # to `{{price}}`.
+  # @return [String]
+  def self.ensure_english_interpolation_values(str, other_str, lng="English")
+    return str unless str.include?("{{") && other_str.include?("{{") && lng === "English"
```
Comment (rgalanakis) [inline lib/suma/i18n.rb:134]: Do we need the `include?` checks?

File: spec/suma/i18n_spec.rb (hunk 2 above)
Comment (rgalanakis) [reply spec/suma/i18n_spec.rb:73]: Similarly, what happens for `{{price}}` and `{{precio cantidad}}` and `{{price}} {{amount}}` and `{{precio}}`?

Comment (rgalanakis) [reply spec/suma/i18n_spec.rb:73]: And add a test to make sure you don't get tripped up by whitespace within the braces.

**Review by rgalanakis — CHANGES_REQUESTED** [review]

File: spec/suma/i18n_spec.rb (hunk 2 above)
Comment (DeeTheDev) [reply spec/suma/i18n_spec.rb:73]: Test would fail due to reversed dynamic values.

Currently, we are only checking for languages that are not English (base_language). This works with our current CSV format so far but maybe there is a better way of checking.

File: spec/suma/i18n_spec.rb [hunk 5]
```diff
@@ -105,6 +105,21 @@
       J
     end
 
+    it "ensures interpolation values are unchanged and removes whitespaces" do
+      path = described_class.strings_path("es")
+      File.write(path, '{"x":1}') # Make sure it gets blown away
+      csv = "Key,Spanish,English\n" \
+            "food:price_times_quantity,{{ precio }} x {{ cantidad }},{{price}} x {{quantity}}"
```
Comment (rgalanakis) [inline spec/suma/i18n_spec.rb:112]: Do `{{cantidad}}` and `{{ quantity }}`, so you test it each way.

File: spec/suma/i18n_spec.rb [hunk 6]
```diff
@@ -105,6 +105,21 @@
       J
     end
 
+    it "ensures interpolation values are unchanged and removes whitespaces" do
+      path = described_class.strings_path("es")
+      File.write(path, '{"x":1}') # Make sure it gets blown away
```
Comment (rgalanakis) [inline spec/suma/i18n_spec.rb:110]: Why do you need this?

File: spec/suma/i18n_spec.rb (hunk 5 above)
Comment (rgalanakis) [inline spec/suma/i18n_spec.rb:112]: Use a different value than 'x' in each string- right now it is unclear whether you're even reading the english vs. spanish string, for example (ie, use `{{ precio }} es {{cantidad}}` and `{{price}} en {{ quanity }}`)

**Review by rgalanakis — CHANGES_REQUESTED** [review]
There is a general problem here where you have something like `{{hello}} {{sir}}!` and. `{{sir}}, {{hello}}!` You'd reverse the string values so it'd be `{{hello}}, {{sir}}!`.

This change is only meant to prevent accidental translation by translators.

Maybe instead of rewriting the string, we should:

- If En and Es have different numbers of placeholders, error, as I don't think we want this to happen yet.
- Grab the strings in the placeholders for each string. If the values are the same (order-independent, ie they may have been reversed intentionally but were not unintentionally translated), use the Es string.
- If the strings in those placeholders are not the same (ie a placeholder was translated), raise an error explaining the problem.
